nexus/src/db/collection_attach.rs

```diff
@@ -0,0 +1,1480 @@
+        // - (for the update) Ensure that only the resource with "resource_id"
+        // is modified.
+        let collection_query = Box::new(
+            collection_query
+                .filter(collection_table().primary_key().eq(collection_id))
+                .filter(Self::CollectionTimeDeletedColumn::default().is_null()),
+        );
+        let resource_query = Box::new(
+            resource_query
+                .filter(resource_table().primary_key().eq(resource_id))
+                .filter(Self::ResourceTimeDeletedColumn::default().is_null())
+                .filter(Self::ResourceCollectionIdColumn::default().is_null()),
+        );
+
+        let update_resource_statement = update
+            .into_boxed()
+            .filter(resource_table().primary_key().eq(resource_id));
+
+        let resource_returning_clause = ResourceType::as_returning();
+        AttachToCollectionStatement {
+            collection_exists_query,
+            resource_exists_query,
+            resource_count_query,
+            collection_query,
+            resource_query,
+            max_attached_resources,
+            update_resource_statement,
+            resource_returning_clause,
+        }
+    }
+}
+
+/// The CTE described in the module docs
+#[must_use = "Queries must be executed"]
+pub struct AttachToCollectionStatement<ResourceType, V, C>
+where
+    ResourceType: Selectable<Pg>,
+    C: DatastoreAttachTarget<ResourceType>,
+{
+    // Query which answers: "Does the collection exist?"
+    collection_exists_query: Box<dyn QueryFragment<Pg> + Send>,
+    // Query which answers: "Does the resource exist?"
+    resource_exists_query: Box<dyn QueryFragment<Pg> + Send>,
+    // Query which answers: "How many resources are associated with the
+    // collection?"
+    resource_count_query: Box<dyn QueryFragment<Pg> + Send>,
+    // A (mostly) user-provided query for validating the collection.
+    collection_query: Box<dyn QueryFragment<Pg> + Send>,
+    // A (mostly) user-provided query for validating the resource.
+    resource_query: Box<dyn QueryFragment<Pg> + Send>,
+    // The maximum number of resources which may be attached to the collection.
+    max_attached_resources: usize,
+
+    // Update statement for the resource.
+    update_resource_statement:
+        BoxedUpdateStatement<'static, Pg, ResourceTable<ResourceType, C>, V>,
+    // Describes what should be returned after UPDATE-ing the resource.
+    resource_returning_clause: AsSelect<ResourceType, Pg>,
+}
+
+impl<ResourceType, V, C> QueryId
+    for AttachToCollectionStatement<ResourceType, V, C>
+where
+    ResourceType: Selectable<Pg>,
+    C: DatastoreAttachTarget<ResourceType>,
+{
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+/// Result of [`AttachToCollectionStatement`] when executed asynchronously
+pub type AsyncAttachToCollectionResult<ResourceType, C> =
+    Result<(C, ResourceType), AttachError<ResourceType, C, PoolError>>;
```

Comment:
This is the result type I expect most users to interact with.

Either:

1. "Here is the collection + resource you updated", or
2. "We did not do the update, and here's an `AttachError` explaining why"